### PR TITLE
Add NRT for the whole client assembly

### LINKED ISSUE #1596: Nullable Reference Types
### Is your feature request related to a problem? Please describe.

Let's add NRT annotations. There are many articles explaining what benefits it bring.

### Describe the solution you'd like

This change is possible to split into 2 changes:
- to mark public API with NRT (breaking change)
- add NRT to internal code

### Describe alternatives you've considered

_No response_

### Additional context

This would be a breaking change, as marking public API with NRT can cause warnings in users projects. If they have TreatWarningsAsErrors, this can cause their projects not built. Therefore, let's consider it as a breaking change.

### LINKED ISSUE #1622: NullReferenceException when setting null Uri into ConnectionFactory
### Describe the bug

It is not clear if ConnectionFactory can have null Uri property. If ConnectionFactory is created via constructor, Uri property returns null. But, ConnectionFactory does not allow explicit set null Uri.

### Reproduction steps

Simple reproduction:
```c#
using RabbitMQ.Client;

var c = new ConnectionFactory();

Console.WriteLine(c.Uri == null); // prints True

c.Uri = null; // throws NullReferenceException
```

### Expected behavior

NullReferenceException is not thrown.

### Additional context

This issue was detected by Roslyn analyzer during work on https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1596.

### PR DESCRIPTION
**Pull request by bollhals**
## Proposed Changes

Fixes #1596
Fixes #1622

Enables Nullable Reference Types in the client assembly

I realize a bit late that you intended to do this in 8.0 ... unfortunately I already was 80% through when I did see it 💯 

You'll have to decide whether it's worth including it or push it after the snap for 7.0

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

How does it work with the PublicAPI thing, does this need adjustment as well?

### CODE REVIEW

File: projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.cs [hunk 1]
```diff
@@ -60,35 +60,29 @@ private Connection InnerConnection
...
             _connectionRecoveryErrorWrapper = new EventingWrapper<ConnectionRecoveryErrorEventArgs>("OnConnectionRecoveryError", onException);
             _consumerTagChangeAfterRecoveryWrapper = new EventingWrapper<ConsumerTagChangedAfterRecoveryEventArgs>("OnConsumerRecovery", onException);
             _queueNameChangedAfterRecoveryWrapper = new EventingWrapper<QueueNameChangedAfterRecoveryEventArgs>("OnQueueRecovery", onException);
 
-            ConnectionShutdown += HandleConnectionShutdown;
+            void onException(Exception exception, string context) =>
+                _innerConnection.OnCallbackException(CallbackExceptionEventArgs.Build(exception, context));
+        }
+
+        internal static async ValueTask<AutorecoveringConnection> CreateAsync(ConnectionConfig config, IEndpointResolver endpoints, CancellationToken cancellationToken)
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.cs:79]: By adjusting the constructor and creating it in this method, the nullablity can be avoided for the Connection.

File: projects/RabbitMQ.Client/client/impl/ChannelBase.cs [hunk 2]
```diff
@@ -1874,7 +1888,7 @@ await tokenRegistration.DisposeAsync()
             }
         }
 
-        private static BasicProperties PopulateActivityAndPropagateTraceId<TProperties>(TProperties basicProperties,
+        private static BasicProperties? PopulateActivityAndPropagateTraceId<TProperties>(TProperties basicProperties,
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/ChannelBase.cs:1891]: I slightly adjusted this avoid creating a new BasicProperty whenever possible and to copy all properties if e.g. the Properties is ReadOnlyProperties

File: projects/RabbitMQ.Client/client/impl/Connection.cs [hunk 3]
```diff
@@ -226,9 +225,6 @@ internal async ValueTask<IConnection> OpenAsync(CancellationToken cancellationTo
 
                 cancellationToken.ThrowIfCancellationRequested();
 
-                await _frameHandler.ConnectAsync(cancellationToken)
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/Connection.cs:229]: The framehandler was actually always connected already

File: projects/RabbitMQ.Client/client/impl/SocketFactory.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using RabbitMQ.Client.Exceptions;
+
+namespace RabbitMQ.Client.client.impl
+{
+    internal static class SocketFactory
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/SocketFactory.cs:10]: I moved these method out here to free up the SocketHandler with this "unrelated" code.

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs [hunk 5]
```diff
@@ -133,107 +126,38 @@ public TimeSpan WriteTimeout
...
             {
-                _writeTimeout = value;
-                SetSocketSendTimout();
+                _socket.Client.SendTimeout = (int)value.TotalMilliseconds;
+                _stream.WriteTimeout = (int)value.TotalMilliseconds;
             }
         }
 
-        public async Task ConnectAsync(CancellationToken cancellationToken)
+        public static async Task<SocketFrameHandler> CreateAsync(AmqpTcpEndpoint amqpTcpEndpoint, Func<AddressFamily, ITcpClient> socketFactory,
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:134]: Similar change here to ensure better nullability

File: projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs [hunk 6]
```diff
@@ -119,8 +112,8 @@ public TimeSpan ReadTimeout
             {
                 try
                 {
-                    _readTimeout = value;
-                    SetSocketReceiveTimeout();
+                    _socket.ReceiveTimeout = value;
+                    _stream.ReadTimeout = (int)value.TotalMilliseconds;
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs:116]: I've set it now also for the stream for consistency

**Comment by bollhals** [comment]
@lukebakken Any idea what I broke with the public api analyzer that it reports internals on the netstandard2.0 target?

**Comment by lukebakken** [comment]
@bollhals somehow this check started to work:

```
dotnet_diagnostic.RS0051.severity = error
```

I disabled it (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md#rs0051-add-internal-types-and-members-to-the-declared-api)

File: projects/RabbitMQ.Client/RabbitMQ.Client.csproj [hunk 7]
```diff
@@ -34,7 +34,8 @@
       https://devblogs.microsoft.com/dotnet/embracing-nullable-reference-types/#what-should-library-authors-do
       Note: only setting language version 8.0 for nullable reference types!
     -->
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>9.0</LangVersion>
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/RabbitMQ.Client.csproj:37]: This gave me concern until I saw what the `npgsql` project uses:

```
C:\Users\lbakken\development\npgsql\npgsql [main ≡]
> git grep -i langversion
Directory.Build.props:4:    <LangVersion>latest</LangVersion>
```

😄

File: projects/Directory.Packages.props [hunk 8]
```diff
@@ -7,6 +7,7 @@
     <PackageVersion Include="Ductus.FluentDocker" Version="2.10.59" />
     <PackageVersion Include="EasyNetQ.Management.Client" Version="2.0.0" />
     <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
+    <PackageVersion Include="Nullable" Version="1.3.1" />
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/Directory.Packages.props:10]: Seems fine to me!

File: projects/RabbitMQ.Client/client/api/IChannel.cs [hunk 9]
```diff
@@ -229,16 +227,14 @@ public interface IChannel : IDisposable
...
 
-#nullable disable
-
         /// <summary>
         /// Configures QoS parameters of the Basic content-class.
         /// </summary>
         /// <param name="prefetchSize">Size of the prefetch in bytes.</param>
         /// <param name="prefetchCount">The prefetch count.</param>
-        /// <param name="global">If set to <c>true</c>, use global prefetch.
+        /// <param name="global">If set to <c>true</c>, use global prefetch.</param>
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/IChannel.cs:235]: Thank you for noticing!

File: projects/RabbitMQ.Client/RabbitMQ.Client.csproj (hunk 7 above)
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/RabbitMQ.Client.csproj:37]: In version 7.0.7 of `npgsql`, `netstandard2.0` is a target framework, and `latest` is still used for `LangVersion`.

File: projects/Directory.Packages.props (hunk 8 above)
[the remark this answers is not in the document]
Comment (lechu445) [reply projects/Directory.Packages.props:10]: https://github.com/manuelroemer/Nullable?tab=readme-ov-file#quickstart says to add Nullable as only development dependency (PrivateAsset=all), therefore it won't be exposed to consumers
```c#
<PackageReference Include="Nullable" Version="<YOUR_VERSION>">
  <PrivateAssets>all</PrivateAssets>
  <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
</PackageReference>
```

File: projects/RabbitMQ.Client/client/api/BasicProperties.cs [hunk 10]
```diff
@@ -85,7 +85,7 @@ public BasicProperties()
         {
         }
 
-        public BasicProperties(ReadOnlyBasicProperties input)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/api/BasicProperties.cs:88]: Is this change related to NRT? If not, consider to extract this breaking change into a separated PR.

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs [hunk 11]
```diff
@@ -647,8 +644,14 @@ private IFrameHandler ConfigureFrameHandler(IFrameHandler fh)
             return fh;
         }
 
-        private void SetUri(Uri uri)
+        private void SetUri(Uri? uri)
         {
+            if (uri is null)
+            {
+                _uri = uri;
+                return;
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:652]: In this case, we may also want to reset parameters that are set based on Uri to the default ones. @lukebakken what is your opinion?

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs [hunk 12]
```diff
@@ -721,7 +724,7 @@ private List<AmqpTcpEndpoint> LocalEndpoints()
             return new List<AmqpTcpEndpoint> { Endpoint };
         }
 
-        private static string EnsureClientProvidedNameLength(string clientProvidedName)
+        private static string? EnsureClientProvidedNameLength(string? clientProvidedName)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:727]: it is possible to add
```c#
[return: NotNullIfNotNull(nameof(clientProvidedName))]
```

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs [hunk 13]
```diff
@@ -71,12 +71,12 @@ public static class EndpointResolverExtensions
...
             }
 
-            if (EqualityComparer<T>.Default.Equals(t, default(T)) && exceptions.Count > 0)
+            if (EqualityComparer<T>.Default.Equals(t!, default!) && exceptions.Count > 0)
             {
                 throw new AggregateException(exceptions);
             }
 
-            return t;
+            return t!;
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: The null-forgiving operator is incorrect here. `t` technically can be null here.

Instead of this, it's better to change the return type from `Task<T>` to `Task<T?>`.

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs [hunk 14]
```diff
@@ -122,7 +123,7 @@ public IRpcContinuation Next()
         /// waiting continuations.
         ///</para>
         ///</remarks>
-        public bool TryPeek<T>(out T continuation) where T : IRpcContinuation
+        public bool TryPeek<T>([MaybeNullWhen(false)] out T continuation) where T : IRpcContinuation
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs:126]:
```suggestion
        public bool TryPeek<T>([NotNullWhen(true)] out T? continuation) where T : IRpcContinuation
```

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs [hunk 15]
```diff
@@ -120,16 +120,16 @@ await TryAuthenticating(options)
...
                 }
             }
-            if (localCertificates != null && localCertificates.Count > 0)
+            if (localCertificates.Count > 0)
             {
                 return localCertificates[0];
             }
 
-            return null;
+            return null!;
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: If this is a valid scenario, the method return type should be `X509Certificate?`. If this is not a valid scenario, and exception should be thrown instead of returning null.

File: projects/RabbitMQ.Client/client/impl/WireFormatting.Write.cs [hunk 16]
```diff
@@ -59,7 +59,7 @@ public static int WriteArray(ref byte destination, IList val)
             return bytesWritten;
         }
 
-        public static int GetArrayByteCount(IList val)
+        public static int GetArrayByteCount(IList? val)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/WireFormatting.Write.cs:62]: There is no usage of this method where null can be passed. Consider to change `List? val` to `List val` and remove unnecessary null check in this method.

File: projects/RabbitMQ.Client/client/api/PublicationAddress.cs [hunk 17]
```diff
@@ -107,7 +107,7 @@ public static PublicationAddress Parse(string uriLikeString)
             return null;
         }
 
-        public static bool TryParse(string uriLikeString, out PublicationAddress result)
+        public static bool TryParse(string? uriLikeString, out PublicationAddress? result)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/api/PublicationAddress.cs:110]:
```suggestion
        public static bool TryParse([NotNullWhen(true)] string? uriLikeString, out PublicationAddress? result)
```

File: projects/RabbitMQ.Client/client/impl/ConsumerDispatching/ConsumerDispatcher.cs [hunk 18]
```diff
@@ -26,26 +25,25 @@ protected override async Task ProcessChannelAsync()
                             try
                             {
                                 IBasicConsumer consumer = work.Consumer;
-                                string? consumerTag = work.ConsumerTag;
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/ConsumerDispatching/ConsumerDispatcher.cs:29]: Consider `string consumerTag = work.ConsumerTag!;`. With this approach you will reduce scope of change in this file.

File: projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs [hunk 19]
```diff
@@ -48,47 +48,48 @@ public static class RabbitMQActivitySource
...
             {
                 return null;
             }
 
-            Activity activity = s_subscriberSource.StartRabbitMQActivity(
+            Activity? activity = s_subscriberSource.StartRabbitMQActivity(
                 UseRoutingKeyAsOperationName ? $"{queue} receive" : "receive",
                 ActivityKind.Consumer);
-            if (activity.IsAllDataRequested)
+            if (activity != null && activity.IsAllDataRequested)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs:92]: I seems activity will never be null here due to implementation detail of StartRabbitMQActivity. Consider to put some assert instead, for example
```c#
Debug.Assert(activity != null, "StartRabbitMQActivity returns null only in case of no active listeners, and this case is already handled above.");
```

File: projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs [hunk 20]
```diff
@@ -107,10 +108,10 @@ internal static Activity ReceiveEmpty(string queue)
             }
 
             // Extract the PropagationContext of the upstream parent from the message headers.
-            Activity activity = s_subscriberSource.StartLinkedRabbitMQActivity(
+            Activity? activity = s_subscriberSource.StartLinkedRabbitMQActivity(
                 UseRoutingKeyAsOperationName ? $"{routingKey} receive" : "receive", ActivityKind.Consumer,
                 ContextExtractor(readOnlyBasicProperties));
-            if (activity.IsAllDataRequested)
+            if (activity != null && activity.IsAllDataRequested)
```
Comment (lechu445) [inline projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs:114]: same as above

File: projects/RabbitMQ.Client/client/api/BasicProperties.cs (hunk 10 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/BasicProperties.cs:88]: Yes, it appears that this change is necessary due to it's use in `PopulateActivityAndPropagateTraceId`

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 11 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:652]: I would prefer that `_uri` and everything related to it is never null. I'll see if that's possible.

File: projects/RabbitMQ.Client/client/api/BasicProperties.cs (hunk 10 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/api/BasicProperties.cs:88]: AFAIK the `ReadOnlyBasicProperties` is anyway new in 7.0, so this should not be problematic

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 12 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:727]: Would be possible but since this is private, I don't see a big benefit in doing so is very limited as none of the usages would benefit from it. Up to @lukebakken.

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs (hunk 13 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: Yes that's true, unfortunately if we'd change the annotation, then that'd mean a whole lot of null checks. So I'd rather throw an exception if there are no endpoints in the resolver. @lukebakken your take

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs:126]: what would be the benefit of this? Except making it worse if T would be a struct.

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: We can't change the definition, as it's from the framework. I didn't touch the original method's way. @lukebakken suppress or throwing

File: projects/RabbitMQ.Client/client/impl/WireFormatting.Write.cs (hunk 16 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/impl/WireFormatting.Write.cs:62]: WireFormatting is formatting according to the protocol, and the list can be null, so even if it isn't used, I wouldn't change it.

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (lechu445) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: It seems that there is missing a nullable annotation in the framework for the result of LocalCertificateSelectionCallback, because it is checked where it is used:
- https://source.dot.net/#System.Net.Security/System/Net/Security/SslStream.Protocol.cs,702
- https://source.dot.net/#System.Net.Security/System/Net/Security/SslStream.Protocol.cs,355

File: projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs (hunk 19 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs:92]: StartRabbitMQActivity can be null though no?

Comment (lechu445) [reply projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs:92]: Technically yes, but in fact StartRabbitMQActivity returns result of `s_subscriberSource.CreateActivity(...)` which returns null only in case when `s_subscriberSource.HasListeners()` returns false. And this case was already handled separately. 
So, it seems the null check that you added does not fix any bug and it's redundant. We can keep it for extra safety of course, if you prefer.

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 11 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs]: @lechu445 @bollhals see my latest commit. Thanks.

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 12 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:764]: I agree there isn't much benefit but I do appreciate the additional "documentation" in the code.  I just added it.

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs (hunk 13 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: Yes, if the `selector` returns `null` there is a serious issue.

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (lechu445) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs:126]: `T` technically can be a struct, but the only usage is with T being ChannelCloseAsyncRpcContinuation class. My suggestion would make sense only if T is restricted to class.
The befit was clarity of contract.
As T can be any type, there is nothing to do for this comment. Resolved.

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: @lechu445 we are constrained by the type signature of the [`LocalCertificateSelectionCallback` delegate](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.localcertificateselectioncallback?view=netstandard-2.0&devlangs=csharp&f1url=%3FappId%3DDev17IDEF1%26l%3DEN-US%26k%3Dk(System.Net.Security.LocalCertificateSelectionCallback)%3Bk(DevLang-csharp)%26rd%3Dtrue).

File: projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs (hunk 20 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs:114]: It's OK as is.

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs]: I've limited TryPeek to `class` and accepted @lechu445's suggestion.

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 12 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:764]: The C# compiler performs static analysis on the code annotated with NRT.

It's not about being public or private code.

That helps in avoiding nullability tests in private code (you should never trust third party inputs).

```csharp
private static string EnsureClientProvidedNameLength(string clientProvidedName)
```

states that `clientProvidedName` is not null (asserted by the developer that wrote the code), which makes the nullability test redundant.

On the other hand, the method may return a `null` reference, which contradicts the method signature.

If you are adding NRT (and you should), you should honor them. Otherwise, you'll end up with incorrect and unverifiable code.

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs (hunk 13 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: > Yes, if the `selector` returns `null` there is a serious issue.

Where is the serious issue? If it's in this method, this method should throw an exception instead of lying that it is returning an value when declaring to only return non-null.

That will cause a `NullReferenceException` to be thrown with no obvious reason to the caller.

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs]: This is the same thing as [`public bool TryGetValue(TAlternateKey key, [MaybeNullWhen(false)] out TValue value)`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Queue.cs,250).

Further more, if `T` is not supposed to be a `struct`, it can be constrained to be a `class`.

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: That looks like an error in the annotation of the [LocalCertificateSelectionCallback delegate](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.localcertificateselectioncallback?view=netstandard-2.0&devlangs=csharp&f1url=%3FappId%3DDev17IDEF1%26l%3DEN-US%26k%3Dk(System.Net.Security.LocalCertificateSelectionCallback)%3Bk(DevLang-csharp)%26rd%3Dtrue).

File: projects/RabbitMQ.Client/RabbitMQ.Client.csproj [hunk 21]
```diff
@@ -34,7 +34,8 @@
       https://devblogs.microsoft.com/dotnet/embracing-nullable-reference-types/#what-should-library-authors-do
       Note: only setting language version 8.0 for nullable reference types!
```
Comment (lechu445) [inline projects/RabbitMQ.Client/RabbitMQ.Client.csproj:35]: is this comment still necessary? If so, should the C# version be updated to 9.0 in this line?

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 12 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:764]: I understand the concept behind, but the NotNullIfNotNull attribute only enhances the definition by telling the compiler the return will not be null if you know the input isn't. 

As this is a private method, it's only called within the same class, where all usages of it are nullable strings as input. So this attribute currently gives no benefit. Hence my comment.

But as @lukebakken said, it can be nice to "document this" / future proof it, which was done now.

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs]: This is why I mostly use `[MaybeNullWhen(false)]` because it's always correct / performant irrespective what type it is.

syntactially AFAIK those are equal (`[MaybeNullWhen(false)]` and `[NotNullWhen(true)]`)

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 11 above)
Comment (lechu445) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs]: Assuming the getter of Uri property won't be called in loop by consumers, this change is fine for me.

**Comment by lukebakken** [comment]
@bollhals @lechu445 - anything else? I plan to merge this by 0900PDT today.

**Comment by lechu445** [comment]
@lukebakken nothing more from my side. LGTM

File: projects/RabbitMQ.Client/client/api/ConnectionFactory.cs (hunk 12 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/api/ConnectionFactory.cs:764]: "if you know" --> "if the compiler can statically prove" 😄

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: Does anyone feel like reporting this in the right github.com/dotnet repo?

File: projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs (hunk 14 above)
Comment (paulomorgado) [reply projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs]: `[MaybeNullWhen(false)]` indicates that the value may be `null` if the method returns `false`, but might not be `null`.

`[NotNullWhen(true)]` indicates that the value is definitely not `null` when the method returns `true`.

All this based on the annotations and developers honoring them.

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs (hunk 13 above)
Comment (lechu445) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: @lukebakken this comment is marked as resolved, although participants agree that logic here should be changed to throw an exception in case of null.

File: projects/RabbitMQ.Client/client/impl/SslHelper.cs (hunk 15 above)
Comment (lechu445) [reply projects/RabbitMQ.Client/client/impl/SslHelper.cs:128]: @paulomorgado that was my first thought too. Then I realized that in https://source.dot.net/#System.Net.Security/System/Net/Security/SslStream.Protocol.cs,702 there is thrown a NotSupportedException in case of null was returned.

File: projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs (hunk 13 above)
Comment (bollhals) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:79]: @lukebakken Should we therefore throw here unconditionally if we get default(T) back? So we can avoid changing the return type?

Comment (lukebakken) [reply projects/RabbitMQ.Client/client/api/IEndpointResolverExtensions.cs:86]: See 6616c4fbac85f98ba0a1bfdc048dee2c7da6bcdc

**Comment by bollhals** [comment]
> @bollhals @lechu445 - anything else? I plan to merge this by 0900PDT today.

With your recent commit, not from my side. 
Hope it is all annotated corectly 🤞

**Comment by lukebakken** [comment]
@bollhals @lechu445 @paulomorgado thank you so much for your insights and contributions here 🥇

**Review by lukebakken — APPROVED** [review]

**Comment by ngbrown** [comment]
I was porting `RabbitMQActivitySource` into my local code from v7 and saw that there were quite a few null reference warnings. Was that code merged after this and not converted?

**Comment by lukebakken** [comment]
@ngbrown it appears to be converted - https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/Impl/RabbitMQActivitySource.cs

**Comment by ngbrown** [comment]
@lukebakken yes, you are correct. I had thought I had copied out the most recent version, but evidently not. Sorry about bothering you.

**Comment by lukebakken** [comment]
@ngbrown no worries at all! If you're testing out the v7 RC candidates, we would appreciate any feedback you have. Thanks!
